design/vela-core/apply-once-only.md

```diff
@@ -0,0 +1,88 @@
+
+If `Apply-Once-Only` is enabled, after creating workload/trait instances, vela-core will skip reconciling them even if they are changed by others (e.g., trait controller, workload controller,etc). 
+Since the create operation is the only one apply operation occurring on workload/trait, we call this mechanism as `Apply-Once-Only`.
+
+## A Motivational Example
+Here is a scenario from production environment to demonstrate how `Apply-Once-Only` works.
+
+```yaml
+apiVersion: core.oam.dev/v1alpha2
+kind: ApplicationConfiguration
+metadata:
+  name: example-appconfig
+spec:
+  components:
+    - componentName: example-component
+---
+apiVersion: core.oam.dev/v1alpha2
+kind: Component
+metadata:
+  name: example-component
+spec:
+  workload:
+    apiVersion: standard.oam.dev/v1alpha1
+    kind: PodSpecWorkload
+    spec:
+      podSpec:
+        containers:
+          - image: sample/app:1.0
+            name: sample-app
+```
+
+After deploying above ApplicationConfiguration, vela-core will create a Deployment with corresponding `PodTemplateSpec`. 
+
+In production env, it's possible to change the Deployment according to particular requirements, e.g., RolloutTrait, AutoscalerTrait,etc.
+Currently, we just use `kubectl` to simulate workload is changed bypass changing the ApplicationConfiguration. 
+Below cmd changes `spec.template.spec.containers[0].image` of the Deployment from `sample/app:1.0` to `sample/app:2.0`.
+```shell
+cat <<EOF | kubectl patch deployment example-deploy --patch
+spec:
+  template:
+    spec:
+      containers:
+      - name: sample-app
+        image: sample/app:2.0
+EOF
+```
+
+Above change will trigger recreate of Pods owned by the Deployment. 
+But vela-core will change it back to `sample/app:1.0` in the following reconciliation soon or late. 
+That's not what we expect in some scenarios.
+
+Instead, we hope vela-core skip reconciling the workload we changed and leave them as what they are now util we change the `spec` of their parent ApplicationConfiguration .
+
+## Goals
+
+Add a switch in vela-core runtime to allow users choose whether always reconcile workload/trait, especially when they are changed by others. 
+
+If the switch is ON, workload/trait will be applied only one time (that is, when they are created).
+After workload/trait is created, vela-core should skip reconcile and apply them EXCEPT below situations:
+
+- The `spec` of ApplicationConfiguration is changed
+- The revision of Component is changed
+
+If the switch is OFF, vela-core will always reconcile and apply workload/trait.
+
+## Implementation
+
+In each round of reconciliation, vela-core compares below Labels & Annotations of existing workload/trait with newly rendered ones before patch-applying or updating-applying. 
+```yaml
+Labels:
+    app.oam.dev/revision
+    app.oam.dev/component
+    app.oam.dev/name
+Annotations:
+    app.oam.dev/generation # generation of ApplicationConfiguration
```

Comment:
Please give a complete example here to explain detail of the implementation

Reply:
Okay